### ISSUE TITLE
False positive when validating type compatibility between metaclass instance and type[T]

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#6285]
The following test case demonstrates the problem.

```python
class CustomMeta1(type):
    ...

def func1(meta: CustomMeta1):
    # This should not generate an error.
    x: type[object] = meta

class CustomMeta2(type):
    ...

class CustomClass2(metaclass=CustomMeta2):
    ...

def func2(meta1: CustomMeta1, meta2: CustomMeta2):
    # This should generate an error.
    x1: type[CustomClass2] = meta1

    # This should not.
    x2: type[CustomClass2] = meta2
```

**Comment by bersbersbers** [comment]
Based on the title, I would guess this issue also fixes the following code, which type-checks in 1.1.333 but not in 1.1.334:

```python
from typing import Callable

class Class:
    __init__: Callable[[str], None]

def fun(class_type: type[Class]):
    class_type(str())
```

```
C:\Build\project\bug.py
  C:\Build\project\bug.py:7:16 - error: Expected 0 positional arguments (reportGeneralTypeIssues)                                                  reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations
```

Otherwise, I can report it separately if required.

**Comment by erictraut** [comment]
@bersbersbers, I can confirm that's a false positive due to a regression introduced in 1.1.334. I can also confirm that it's fixed in the latest main.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.335, which I just published. It will also be included in a future release of pylance.
